src/Client.php

```diff
@@ -272,7 +273,126 @@ public function getAsset($assetId) {
       ['headers' => $this->getDefaultHeaders()]
     );
 
-    return (string) $response->getBody();
+    $asset = Asset::fromJson((string) $response->getBody());
+
+    return $asset;
+  }
+
+  /**
+   * Gets presigned url from AWS S3.
+   *
+   * @param array $file_data
+   *   The file data required by Webdam
+   *   (filesize, folder_id, filename, contenttype).
+   *
+   * @return mixed
+   *   Presigned url needed for next step + PID.
+   */
+  protected function getPresignUrl(array $file_data) {
+    $this->checkAuth();
+
+    $response = $this->client->request(
+      "GET",
+      $this->baseUrl . '/ws/awss3/generateupload',
+      [
+        'headers' => $this->getDefaultHeaders(),
+        'query' => $file_data,
+      ]
+    );
+    return json_decode($response->getBody());
+  }
+
+  /**
+   * Uploads file to Webdam AWS S3.
+   *
+   * @param mixed $presignedUrl
+   *   The presigned URL we got in previous step from AWS.
+   * @param string $file_uri
+   *   The file URI.
+   * @param string $file_type
+   *   The File Content Type.
+   * @return array
+   *   Response Status 100 / 200
+   */
+  protected function uploadPresigned($presignedUrl, $file_uri, $file_type) {
+    $this->checkAuth();
+
+    $response = $this->client->request(
+      "PUT",
+      $presignedUrl, [
+        'headers' => ['Content-Type' => $file_type],
+        'body' => $file_uri,
+      ]);
+    return [
+      'status' => json_decode($response->getStatusCode(), TRUE),
+    ];
+
+  }
+
+  /**
+   * Confirms the upload to Webdam.
+   *
+   * @param string $pid
+   *   The Process ID we got in first step.
+   *
+   * @return string
+   *   The uploaded/edited asset ID.
+   */
+  protected function uploadConfirmed($pid) {
+    $this->checkAuth();
+
+    $response = $this->client->request(
+      "PUT",
+      $this->baseUrl . '/ws/awss3/finishupload/' . $pid,
+      ['headers' => $this->getDefaultHeaders()]
+    );
+
+    return (string) json_decode($response->getBody(), TRUE)['id'];
+
+  }
+
+  /**
+   * Uploads Assets to Webdam using the previously defined methods.
+   *
+   * @param array $file_data
+   *   The file data required by Webdam.
+   * @param int $folderID
+   *   The Webdam folder ID.
+   *
+   * @return array
+   *   Webdam response (processId, presign url, status, asset id).
+   */
+  public function uploadAsset(array $file_data, $folderID = NULL) {
```

Comment:
I'd still like to see the keys of the $file_data array broken out into separate arguments so that they can be documented. I have no idea what should be in this array looking in from the outside - let's just make everything separate arguments to the uploadAsset() method.